src/gtest/test_joinsplit.cpp

```diff
@@ -0,0 +1,213 @@
+#include <gtest/gtest.h>
+
+#include "utilstrencodings.h"
+
+#include "zcash/prf.h"
+
+#include "zcash/JoinSplit.hpp"
+#include "zcash/Note.hpp"
+#include "zcash/NoteEncryption.hpp"
+#include "zcash/IncrementalMerkleTree.hpp"
+
+using namespace libzcash;
+
+void test_full_api(ZCJoinSplit* js)
+{
+    // The recipient's information.
+    SpendingKey recipient_key = SpendingKey::random();
+    PaymentAddress recipient_addr = recipient_key.address();
+
+    // Create the commitment tree
+    ZCIncrementalMerkleTree tree;
+
+    // Set up a JoinSplit description
+    uint256 ephemeralKey;
+    uint256 randomSeed;
+    uint64_t vpub_old = 10;
+    uint64_t vpub_new = 0;
+    uint256 pubKeyHash = random_uint256();
+    boost::array<uint256, 2> hmacs;
+    boost::array<uint256, 2> nullifiers;
+    boost::array<uint256, 2> commitments;
+    uint256 rt = tree.root();
+    boost::array<ZCNoteEncryption::Ciphertext, 2> ciphertexts;
+    std::string proof;
+
+    {
+        boost::array<JSInput, 2> inputs = {
+            JSInput(), // dummy input
+            JSInput() // dummy input
+        };
+
+        boost::array<JSOutput, 2> outputs = {
+            JSOutput(recipient_addr, 10),
+            JSOutput() // dummy output
+        };
+
+        boost::array<Note, 2> output_notes;
+
+        // Perform the proof
+        proof = js->prove(
+            inputs,
+            outputs,
+            output_notes,
+            ciphertexts,
+            ephemeralKey,
+            pubKeyHash,
+            randomSeed,
+            hmacs,
+            nullifiers,
+            commitments,
+            vpub_old,
+            vpub_new,
+            rt
```

Comment:
I think I've mentioned this before, but it would be much clearer if the output parameters were prefixed with ``out_``.